### PR TITLE
sql: disable Streamer in some illegal cases

### LINKED ISSUE #122274: roachtest: schemachange/mixed-versions failed [attempting to append refresh spans after the tracked timestamp has moved forward]
roachtest.schemachange/mixed-versions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14807939?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14807939?buildTab=artifacts#/schemachange/mixed-versions) on release-23.2 @ [e61e140a75b2f7dde9f5410c1d90e557d009c437](https://github.com/cockroachdb/cockroach/commits/e61e140a75b2f7dde9f5410c1d90e557d009c437):

```
(mixedversion.go:592).Run: mixed-version test failure while running step 17 (run "run schemachange workload and validation in mixed version"): full command output in run_062406.146932602_n4_COCKROACHRANDOMSEED1.log: COMMAND_PROBLEM: exit status 1
test artifacts and logs in: /artifacts/schemachange/mixed-versions/run_1
```

Parameters:
 - <code>ROACHTEST_arch=amd64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_metamorphicBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-14807939/schemachange-mixed-versions/1712902763107/1712903188364)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #121821 roachtest: schemachange/mixed-versions failed [issue renaming and using a function in the same txn] [C-test-failure GA-blocker O-roachtest O-robot P-1 T-sql-foundations branch-master]
- #121411 roachtest: schemachange/mixed-versions failed [C-test-failure O-roachtest O-robot P-1 T-sql-foundations branch-release-24.1]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*schemachange/mixed-versions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-37787

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, it was possible to run into an internal error due to the usage of the Streamer API in the main query while part of the query (e.g. a cast to an Oid type) uses the Internal Executor with the RootTxn. This is illegal and might trigger `attempting to append refresh spans after the tracked timestamp has moved forward` error. We fix this by disabling the Streamer whenever usage of DistSQL was prohibited.

This seems reasonable because both DistSQL and Streamer provide concurrency (but in different forms) and require usage of the LeafTxns, so if we couldn't distribute the plan, then we likely can't use the Streamer either. Note that this check might be stricter than necessary (DistSQL could be prohibited due to some serialization issues that Streamer wouldn't be affected by). Additionally, if the plan is not distributed because `distsql=off` is used, then we don't know whether it would have been prohibited or not, so we might not disable the Streamer when we should.

Still, this seems like a reasonable improvement.

Fixes: #122274.

Release note (bug fix): CockroachDB could previously run into `attempting to append refresh spans after the tracked timestamp has moved forward` internal error in some edge cases, and this is now fixed. The bug has been present since 22.2 version.

### CODE REVIEW

**Review by DrewKimball — APPROVED** [review]
:lgtm: Nice fix. Do you think there's any easy opportunity to allow the streamer in common cases where distsql is disallowed for reasons other than concurrency?

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123111#-:-NwOGiBn0oJK7Qu-mmqR:b-bsadtv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

**Comment by yuzefovich** [comment]
Effectively, what this patch does is disabling the streamer whenever `checkExprForDistSQL` returns an error (because other couple of cases which get `distSQLProhibitedErr` are with wrapped `planNode`s and locking, and for both we already disable the streamer). Looking at `checkExprForDistSQL` I only see a few subcases that are about serialization AFAICT: subquery returning an Oid type, having `DArray` with untyped tuples, having `DTuple` that is untyped, and these seem rare enough to not special case them.

There are relatively large buckets in `checkExprForDistSQL` where it's not clear whether it'd be problematic to use the streamer or not:
- distSQL blocklist for builtins - for some we do need to disable the streamer (e.g. where we call `ParseDOid`), for others it might not be needed
- routines - since the body might have arbitrary statements, we might need to disable the streamer, but it seems hard to determine that
- having `DOid` or a cast to it - need to disable.

In short, it seems reasonable to me - at least for now - to disable streamer whenever distsql is explicitly prohibited.

TFTR!

bors r+
